Always refresh ring on topology changes and reconnections

joao-r-reis opened the pull request:
Currently, there are some cases where gocql does not refresh the ring during a control connection reconnection. If topology changes are happening on the cluster during the reconnection, some `NEW_NODE` and `REMOVED_NODE` protocol events might be missed. 

To fix this, every successful reconnection on the control connection should trigger a ring refresh operation. This PR adds that.

Also, this PR fixes an issue where `NEW_NODE` events can be overwritten by `UP` events and cause certain nodes to not be added to the ring. This is an issue already described in #1669 but the fix proposed by this PR is a bit different than one outlined on that PR. This PR causes any topology event to trigger a ring refresh regardless of how many nodes were removed or added. This eliminates the need to have duplicated logic around adding and removing nodes in both `ringDescriber` and `Session`.

**Note**: this is not an implementation of #1681.

martin-sucha commented:
Hi! Thanks for the pull request.

Did you mean to put removal of `addNewNode` to the [remove unused code](https://github.com/gocql/gocql/pull/1680/commits/fff05239a6593369c4da5cd90ec6970067587e6a) commit instead of `handleNewNode` and `handleRemovedNode`?

Could you please update the description of commit https://github.com/gocql/gocql/pull/1680/commits/326072be085e740a48f545521301691f99261e2d to what the description of the merge request says? `refresh ring improvements` is a bit vague. The commit title could be something like  `Always refresh ring on topology change`. That should make it easier to find the commit in commit log and see the more details with `git blame`.

It seems gocql currently does not de-bounce the ring refresh, so always refreshing the ring could theoretically cause a lot of loads if events come in quick succession. I suspect this is not a problem in practice, but we can still add the de-bounce (maybe in a separate pull request?) to guard against that.

Another thing that I noticed while reviewing the code is that `refreshRing` loads data from `system.peers`, but seems to not use data from `system.local`. That happens only in `addNewNode` resp. `ringDescriber.getHostInfo`, which is not used in current master. We should read from system.local in ringDescriber.GetHosts as well.

Otherwise this looks good to me.

It would be great if you could add a test to make sure that we don't introduce the same issues in the future.

joao-r-reis commented:
Thanks for the quick review. I've changed the commits a bit so they should be clearer now.

> Did you mean to put removal of addNewNode to the [remove unused code](https://github.com/gocql/gocql/pull/1680/commits/fff05239a6593369c4da5cd90ec6970067587e6a) commit instead of handleNewNode and handleRemovedNode?

I removed `addNewNode` initially and then I found out that `handleNewNode` and `handleRemovedNode` were also unused so I removed these 2 later as well (now they are in the same commit).

> It seems gocql currently does not de-bounce the ring refresh, so always refreshing the ring could theoretically cause a lot of loads if events come in quick succession. I suspect this is not a problem in practice, but we can still add the de-bounce (maybe in a separate pull request?) to guard against that.

The topology change events are being debounced but I agree that it would probably be a good idea to debounce the ring refresh operations themselves. I'm happy to address this in a subsequent PR or I can just create a new issue so another contributor can pick it up, let me know what you prefer.

> Another thing that I noticed while reviewing the code is that refreshRing loads data from system.peers, but seems to not use data from system.local. That happens only in addNewNode resp. ringDescriber.getHostInfo, which is not used in current master. We should read from system.local in ringDescriber.GetHosts as well.

I'll start working on this and I'll also look into writing a test for this.

martin-sucha commented:
> > Did you mean to put removal of addNewNode to the [remove unused code](https://github.com/gocql/gocql/pull/1680/commits/fff05239a6593369c4da5cd90ec6970067587e6a) commit instead of handleNewNode and handleRemovedNode?
> 
> I removed `addNewNode` initially and then I found out that `handleNewNode` and `handleRemovedNode` were also unused so I removed these 2 later as well (now they are in the same commit).

Techically, `handleNewNode` and `handleRemovedNode` are used until https://github.com/gocql/gocql/pull/1680/commits/7b506bcb0421440ad23c02b53d830b35c7509b60 so the removal should be part of that commit.

> 
> > It seems gocql currently does not de-bounce the ring refresh, so always refreshing the ring could theoretically cause a lot of loads if events come in quick succession. I suspect this is not a problem in practice, but we can still add the de-bounce (maybe in a separate pull request?) to guard against that.
> 
> The topology change events are being debounced but I agree that it would probably be a good idea to debounce the ring refresh operations themselves. I'm happy to address this in a subsequent PR or I can just create a new issue so another contributor can pick it up, let me know what you prefer.

Feel free to send a pull request, that would be great.

> > Another thing that I noticed while reviewing the code is that `refreshRing` loads data from `system.peers`, but seems to not use data from `system.local`. That happens only in `addNewNode` resp. `ringDescriber.getHostInfo`, which is not used in current master. We should read from system.local in ringDescriber.GetHosts as well.
> 
> I'll start working on this and I'll also look into writing a test for this.

Thanks!

joao-r-reis commented:
I've rebased my branch and pushed commits to make `GetHosts()` retrieve the local host info from `system.local` and to remove `getHostInfo()` which was only used by `addNewNode()` and this function was removed in one of my earlier commits because it was also unused.

For the debouncer, I would think having an event debouncer would be enough and I see that `gocql` already has one, is there a need to have two debouncers? What would be the goal of the ring refresh debouncer?

StevenLacerda commented:
@martin-sucha Can you take a look at the merge request? This should help tremendously with K8's environments. Thanks!

joao-r-reis commented:
I've implemented a debouncer, can you take a look and see if this is aligned with what you were thinking? I'm going to look into tests now.

joao-r-reis commented on host_source.go:
```diff
@@ -721,3 +732,151 @@ func (r *ringDescriber) refreshRing() error {
+}
+
+func (d *ringRefreshDebouncer) debounce() {
+	d.timer.Reset(ringRefreshDebounceTime)
```
I might add a lock here as well, not sure if concurrent `Reset` and `Stop` calls can be problematic here so should probably lock just in case

martin-sucha commented:
> I've added unit tests but I don't see a way to simulate a topology change for an integration test, does the current test code support writing an integration test that could test whether the control connection updates the topology when it reconnects?

Yeah, the current integration tests use ccm and it would probably be hard to write an integration test that changes topology. We should probably use Go code directly instead of ccm to setup the cluster. That is a bigger chunk of work though. I've opened https://github.com/gocql/gocql/issues/1686 to track that.

> Oh I've also changed the host_source_test.go tests to run with the unit flag instead of cassandra flag since this file seems to contain unit tests only.

Nice catch. Thanks!

martin-sucha commented on host_source.go:
```diff
@@ -721,3 +732,160 @@ func (r *ringDescriber) refreshRing() error {
+	d := &refreshDebouncer{
+		stopped:      false,
+		broadcaster:  nil,
+		refreshNowCh: make(chan struct{}),
```
It seems that refreshNowCh should have capacity one instead of zero. The sender does non-blocking send, so with capacity zero, there is a chance that the refresh signal is lost when there is no receiver (for example, while refreshFn is running or while errorBroadcaster.broadcast is running).

joao-r-reis replied:
Oh nice catch, I missed this. I've changed this to have a capacity of 1.

martin-sucha reviewed (changes requested):
Thank you for preparing the changes! Looks good to me in general, I've added some inline comments.

joao-r-reis commented:
Managed to add a ccm test that tests whether the control connection refreshes the ring on reconnect. I used `DisableTopologyEvents` config and a custom `HostFilter` to avoid using real topology changes.

I'll take a look at your review feedback now, thanks.

joao-r-reis commented on host_source.go:
```diff
@@ -721,3 +732,160 @@ func (r *ringDescriber) refreshRing() error {
+		return
+	}
+	d.stopped = true
+	d.quit <- struct{}{}
```
I've tried to keep the `stop()` behavior as close as possible to the `eventDebouncer` but I didn't copy the comment that is on that one. I've copied that comment now so it is clear that we are syncing with the flusher goroutine.

martin-sucha commented on host_source.go:
```diff
@@ -721,3 +732,160 @@ func (r *ringDescriber) refreshRing() error {
+		return
+	}
+	d.stopped = true
+	d.quit <- struct{}{} // sync with flusher
```
It seems we need to unlock `d.mu` before waiting for flusher to exit, otherwise there could be a deadlock similar to #1687. If refresh is in progress, `stop()` waits until flusher completes, which waits for refresh to complete first. The refresh might take a long time and something could call `refreshNow` or `debounce` during that time, which would be blocked on acquiring `d.mu`.

martin-sucha reviewed (changes requested):
Thanks! I've noticed one more thing, added inline comment.

StevenLacerda commented:
@martin-sucha @joao-r-reis Is this ready to be reviewed and merged? Really looking forward to getting this out as it will help several of our customers.

joao-r-reis commented:
I'm waiting on a review and ready to implement any changes that are suggested. I've addressed all comments of the previous reviews.

martin-sucha commented:
Looks good to me. Thank you!

martin-sucha commented:
It seems there is a data race in the ccm test. @joao-r-reis could you please check https://github.com/gocql/gocql/pull/1691 ?

StevenLacerda commented:
@joao-r-reis @martin-sucha Any updates? What's needed at this point?

joao-r-reis commented:
> @joao-r-reis @martin-sucha Any updates? What's needed at this point?

This has already been released with v1.4.0 AFAIK
